iothub_client/src/iothubtransport_amqp_messenger.c

```diff
@@ -86,9 +87,24 @@ typedef struct MESSAGE_SEND_CONTEXT_TAG
     void* user_context;
 
     PROCESS_MESSAGE_COMPLETED_CALLBACK on_process_message_completed_callback;
+
+    // Handle to the async operation instance returned
+    // by messagesender_send_async().
+    ASYNC_OPERATION_HANDLE async_operation;
 } MESSAGE_SEND_CONTEXT;
 
+static void destroy_message_send_context(MESSAGE_SEND_CONTEXT* context)
+{
+    if (context->async_operation != NULL)
+    {
+        (void)async_operation_cancel(context->async_operation);
+    }
+
+    free(context);
+}
 
+#define mark_message_send_context_completed(context) \
```

Comment:
Will this ever cause a problem in destroy_message_send_context that context->async_operation is null, but hasn't been cancelled?

Reply:
If context->async_operation is NULL, it means the send complete callback was called, so there is nothing to cancel anymore. mark_message_send_context_completed shall only be called from within that send complete callback.